api/src/main/java/org/infinispan/api/Infinispan.java

```diff
@@ -1,50 +1,104 @@
-               .newInstance(config);
-      } catch (Exception e) {
+public interface Infinispan extends AutoCloseable {
+   /**
+    * Creates and starts the Infinispan manager object
+    *
+    * Calling create with the same URI multiple times should return the same object. Ref count the object.
+    *
+    * <ul>
+    *   <li><tt>file:///.../infinispan.xml</tt> Embedded Infinispan configured via XML/JSON/Yaml file</li>
+    *   <li><tt>classpath:///.../infinispan.xml</tt> Embedded Infinispan configured via XML/JSON/Yaml classpath resource</li>
+    *   <li><tt>hotrod[s]://[username[:password]@]host:port[,host2:port]?property=value[&property=value]</tt> </li>
+    * </ul>
+    *
+    * @param uri one of the supported Infinispan URIs:
+    * @return an
+    */
+   static Infinispan create(URI uri) {
+      Iterator<Factory> factories = ServiceLoader.load(Factory.class, Factory.class.getClassLoader()).iterator();
+      while (factories.hasNext()) {
+         Factory factory = factories.next();
+         Infinispan instance = factory.create(uri);
+         if (instance != null) {
+            return instance;
+         }
       }
-      return null;
+      throw new InfinispanConfigurationException("No factory to handle URI " + uri);
    }
 
-   static Infinispan newEmbedded(EmbeddedConfig config) {
-      throw new UnsupportedOperationException("Embedded mode not yet supported");
+   static Infinispan create(String uri) {
+      return create(URI.create(uri));
    }
 
+   static Infinispan create(Configuration configuration) {
+      Iterator<Factory> factories = ServiceLoader.load(Factory.class, Factory.class.getClassLoader()).iterator();
+      while (factories.hasNext()) {
+         Factory factory = factories.next();
+         Infinispan instance = factory.create(configuration);
+         if (instance != null) {
+            return instance;
+         }
+      }
+      throw new InfinispanConfigurationException("No factory to handle configuration " + configuration);
+   }
+
+   /**
+    * Returns a synchronous version of the Infinispan API
+    *
+    * @return
+    */
+   SyncContainer sync();
+
    /**
-    * Gets the {@link KeyValueStore} by name.
-    * <p>
-    * If the store does not exist, creates a {@link KeyValueStore} with the given config
+    * Returns an asynchronous version of the Infinispan API
     *
-    * @param name, name of the store
-    * @return {@link KeyValueStore}
+    * @return
     */
-   <K, V> CompletionStage<KeyValueStore<K, V>> getKeyValueStore(String name, KeyValueStoreConfig config);
+   AsyncContainer async();
 
    /**
-    * Stops Infinispan
+    * Returns a mutiny version of the Infinispan API
     *
-    * @return {@link CompletionStage}
+    * @return
     */
-   CompletionStage<Void> stop();
+   MutinyContainer mutiny();
+
+   /**
+    * Closes the instance, releasing all allocated resources (thread pools, open files, etc)
+    */
+   @Override
+   void close();
```

Comment:
I assume we're overriding here so that we can suppress the `throws Exception`?

Shouldn't the user be aware of any exceptions on close?

Reply:
They will be runtime exceptions

Reply:
Hmm for me this is a good use-case for checked exceptions, especially considering the alternative is effectively changing the behaviour of a well known and understood `AutoCloseable` interface.

Reply:
Do we want this method to be blocking as well? I think we need an asynchronous close version as well.